R/utils.R

```diff
@@ -235,45 +255,94 @@ assert_sufficient_f_args <- function(f, ...) {
 #'
 #' @noRd
 as_slide_computation <- function(x,
+                        before,
+                        calc_ref_time_value = FALSE,
```

Comment:
Is it possible to do the `f_wrapper` stuff only in `epi_slide` on the result of `as_slide_computation`, and not have to pass `before` and `calc_ref_time_value` at all?

Regarding name collisions, I'm more worried about `x`; could things go wrong here if the user wants to make a `x` column or use an arg called `x`?  Maybe we could rename it to `f` or `.f`?  Or perhaps we can avoid dots passing here at all; could we pass instead the number of dots, or the names of the dots?  (Or is there some way to pack up dots?  Maybe `enquos`?  Otherwise this doesn't look trivial without some compromise.  `rlang::dots_list` evaluates all the dots (breaks computation-internal laziness / tidyeval).  Packing directly an environment may not work with unnamed args.  If all we need is to be able to forward dots but not inspect things about them, then maybe we could use a wrapper function that will forward dots to a function.  There might be something with `call_match` to make things work, not sure.)

Reply:
(Another idea on name collisions... we could make `as_slide_computation` into an appropriately-named higher-order function `function(x, before, etc.) function(...) { <body> }` to hopefully avoid any conflicts.)

Reply:
> Is it possible to do the f_wrapper stuff only in epi_slide on the result of as_slide_computation, and not have to pass before and calc_ref_time_value at all? [oh, I'm guessing this involves some extra function dispatch overhead in the epi_slide tidyeval case?]

Great idea, that makes everything much simpler. Yes, it would add function dispatch time, but looks like it's not worth worrying about. Using [this same test code](https://github.com/cmu-delphi/epiprocess/pull/318#issuecomment-1568835601), median runtime is 56.4 ms vs 56.6 ms for the version at https://github.com/cmu-delphi/epiprocess/pull/337/commits/a18be4992cb9b73285ebf0543ac58ccfaeae651c using the `calc_ref_time_value` flag vs [~56 ms](https://github.com/cmu-delphi/epiprocess/pull/318#discussion_r1220403431) for the version at https://github.com/cmu-delphi/epiprocess/pull/318/commits/9ccfe915f7d348528465ce105d6bc8d7fa733a9c, so no appreciable difference.

Reply:
> Regarding name collisions, ... Maybe we could rename it to f or .f?

I went ahead and renamed to `.f` as a start.

The dots are only used in the `assert_sufficient_f_args` check in the function case, which needs both names and length of `...` so passing only those attributes could work. We could put `assert_sufficient_f_args` back into `epi[x]_slide` to avoid the issue entirely.

I'm not sure what you mean about a "higher-order function". Could you expand on that.

Reply:
There are various places (`epi[x]_slide` + internals, `assert_sufficient_f_args`, `as_slide_computation`) that we pass user-provided dots to that could cause name problems. Would it make sense to make a helper function that we could use in all those places to explicitly checks for name collisions between dots and a function? Like
```
if (any(names(rlang::dots_list(...)) %in% rlang::fn_fmls_names())) {
  Abort("<error>")
}
```

Reply:
Oh, and also the prefix `.`.

Reply:
I changed `as_slide_computation` to take only `f` and `...` args so that this is true:

> if ... all our non-`...` parameter names are also parameter names of epi_slide and epix_slide ... the collision happens immediately and hopefully? more clearly, rather than confusingly in some helper function

`as_slide_computation` is called from `epi[x]_slide`, both of which have an `f` arg.